### ISSUE TITLE
"is" operator

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#2976]
In order to be able to properly check at runtime abstract types, a "is" operator is needed.
See #2617 for the whole discussion

**Comment by Simn** [comment]
Mind if we implement this for 3.3? The recent Std.is discussion made me remember how much I want a proper is operator.

**Comment by nadako** [comment]
Yes please!!!

**Comment by nadako** [comment]
I think we also need a warning when `Std.is` second argument is not a `TTypeExpr` to prepare users for the migration.

**Comment by Simn** [comment]
I'll implement this as `(expr is type_path)` with the mandatory parentheses for now. We can do this without making `is` a keyword.

By the way, safe casts allow parsing a complex type instead of a type path. Does this make any sense? It seems like the only things we allow there are type paths anyway.

**Comment by ncannasse** [comment]
macros ?

**Comment by ViktorLajos** [comment]
I found an issue with the 'is' operator, and Array type checking!

```haxe
public function new()
{
    var a = new Array<Float>();
    isTest(a);
} 

public function isTest(test:Dynamic):Void
{
    if((test is String))
    {
        trace("test is String");
    }else if((test is Array<String>))
    {
        trace("test is  Array<String>");
    }else
    {
        trace("test is unkonwn");
    }
}
```

and it traces the next:

>  [trace] Main.hx:26: test is  Array<String>

**Comment by ncannasse** [comment]
@Simn we shouldn't allow type parameters other than Dynamic (unless generic / macrotype)

**Comment by Simn** [comment]
Technically yes, but at the moment it's merely a translation to `Std.is` so we can't make it fail during typing.

**Comment by kobi2187** [comment]
on latest haxe, 'is' only works when surrounded by (). an if statement needs double parentheses.
`if ((a is Int))`

**Comment by yar3333** [comment]
I think mandatory parentheses looks ugly. May be we can avoid that? I think this is a small breaking change.

**Comment by Simn** [comment]
There was some problem that requires the parentheses. I remember thinking that we would have to make `is` a keyword in order to lose them, but now I'm not quite sure why that would be the case.

Still, I think `is` should be a keyword regardless.

**Comment by Simn** [comment]
... and the main problem with that is that `Std.is` wouldn't be possible anymore.

**Comment by kevinresol** [comment]
> ... and the main problem with that is that Std.is wouldn't be possible anymore.

I still don't quite get that why that would be a problem. Given that it is unambiguous that "is" is a field access, because it is after a dot.

See: https://github.com/HaxeFoundation/haxe-evolution/issues/22

**Comment by Simn** [comment]
I'm not sure what to do here. We can't make `is` a keyword, so the only option would be to treat it as a pseudo-keyword in `expr_next`. That might lead to syntactic issues though.

**Comment by RealyUniqueName** [comment]
Maybe start with deprecating `Std.is()` (and warn on all `is` symbols) and introducing something like `Std.instanceof` instead. And then a few versions later remove `Std.is()` and add `is` keyword.

**Comment by ncannasse** [comment]
I'm a bit not sure also, because I don't like much having `Std.is(v,e)` (or its replacement) taking an expression value and `v is T` taking a Type... which can in many cases also be an expression value - unless it has type parameters that we can rarely check anyway.

Ideally we would make the `is` keyword accept both CompleType and Expr so we can do both `x is Abstract<String>` and `x is getType()`,  so Std.is will be a deprecated shortcut of `is` operator using an expression but I'm not sure it will work well wrt parsing.

Any other insights regarding the duality check with type vs expression ?
It's mostly a parsing problem because in both cases we need an actual value to check against at runtime.

**Comment by Simn** [comment]
I can't see that working parsing-wise. There are plenty of token sequences that are accepted as both type and expression, and there are even more that are partially accepted but then fail.

The best solution here is to rename `Std.is` like @RealyUniqueName suggests. That way we can use it as the expression-version, while we can use `expr is Type` for the type-version.

I propose the following approach:

### 4.0

* Parse `expr is Type` and treat `is` as a pseudo-keyword
* Introduce `Std.isInstance`
* Deprecate `Std.is`
* (maybe) Detect usages of `is` and give warnings about it

### 4.1

* Make `is` a keyword.
* Remove `Std.is`

**Comment by jdonaldson** [comment]
Should we try and address some of the semantics of "is" as well?   

```haxe
    var k : Float = 1.0;
    trace(Std.is(k,Int)); // true
```
Platforms are not going to have 1-1 mappings of value types to runtime types.  I'm wondering if it makes sense to make "is" operate only on certain types (of types), and then handle the rest with special one-off methods (e.g.  `isNumeric()`).

**Comment by ncannasse** [comment]
We have to do something about the naming, because I'm not much happy to have both `Std.instanceOf` and `Std.isInstance` , or am I the only one ?

**Comment by nadako** [comment]
The more proper name would be something along `Std.isOfType`, because the word "instance" is kind of reserved by class instances.

**Comment by ncannasse** [comment]
@jdonaldson in that particular case I favor the crossplatform behavior over the platform specific meaning of a runtime type, I'm not sure if it's the best or not.

**Comment by Simn** [comment]
I don't really know what to say when the person who added `Std.instance` is not happy with naming. x)

I also don't have a better suggestion...

**Comment by jdonaldson** [comment]
Another observation: most (all?) of the problematic types here are abstract.  In general, since abstract types don't always have a distinct runtime signature, they'll be undetectable in many cases.  I was going to make a small note in the `Std.is` documentation to be wary of type checks on certain platforms.  

I've slept on this a few times, and I keep going back and forth between "we can just leave this as undefined behavior" and "this is a commonly used method that (if left unchecked) would generate code that is not cross platform".

Just now I'm leaning towards moving the method to the `Reflect` class, so that at least people will know that it carries the same caveats as the rest of the methods there.

**Comment by jdonaldson** [comment]
Linking in the "strong belgian ale" related issue : https://github.com/HaxeFoundation/haxe/issues/5202

TLDR : `Std.is` has different behavior on the other platforms.

(Note: I bravely picked up this bug voluntarily after a night of drinking strong belgian ale)

**Comment by jdonaldson** [comment]
Linking the `Std.is(null,T)` specification : https://github.com/HaxeFoundation/haxe/issues/7115

TLDR : NC gives a spec for how to handle null checks with this method.

**Comment by jdonaldson** [comment]
Linking to `Std.is` issue on java : https://github.com/HaxeFoundation/haxe/issues/5168

TLDR : Java will throw an error (rather than returning false) when performing `Std.is` on incompatible types.

**Comment by ncannasse** [comment]
We need some decision on naming.

Maybe we could have `Std.isInstance` and  `Std.toInstance` but they both have different behavior, the first one can handle types such as Int, Array or interfaces while the second one is specified for class only.
(BTW I would like to add support for interfaces in Std.instance but that's another story).

So it seems they cannot use the same vocabulary. 

Any other suggestion?

**Comment by Simn** [comment]
I think we have to forego this for the time being because I don't see us coming up with a satisfactory solution here...

**Comment by ncannasse** [comment]
I agree

**Comment by kevinresol** [comment]
When `is` becomes an operator does that mean it can be customized in abstracts?

e.g.
```haxe
var isInt64 = x is Int64;

abstract Int64 {
  @:op(is) static function isInt64(v:Dynamic):Bool;
}
```

**Comment by RealyUniqueName** [comment]
So, how about introducing `Std.isOfType()` and deprecating `Std.is`?
`isOfType` would work great with static extensions: `value.isOfType(MyType)`.
As for `Std.is` we can add a deprecation notice in a doc block for 4.1 and add `@:deprecated` in 4.2.
I'm going to wait a week for responses and then do that voluntary )

**Comment by farteryhr** [comment]
idea: separate the one for primitive types (like using js:typeof, lua:type(x), matches a finite set, do simple comparison, summarize to "object" "function" for things with complex types it can't handle) and the one for objects (like using instanceof, plus proper Array handling, and more complicated type-type matching, always return false when the lhs operand is primitive value(?)).

by instinct, the mixed way feels like makeing both pain and performance loss. not some rigid point though... just feeling..

**Comment by RealyUniqueName** [comment]
I don't think it's a significant performance loss. And it's a bad idea to use any of that in a performance critical code anyway.

**Comment by RealyUniqueName** [comment]
> add @:deprecated in 4.2

The time has come. Or we can postpone this to 4.3.
Any opinions?

**Comment by Gama11** [comment]
I don't think waiting until 4.3 would help much, so I'm in favor of adding it now.

**Comment by back2dos** [comment]
Is promoting `is` to a keyword really necessary?

It does cause the invention of new names for [native APIs](https://github.com/HaxeFoundation/haxe/commit/3ba716717c3839243cc816c6cdc7fb44f6fd830f#diff-5e8cfd8c32a6ef5ca5cfdb720380f17fR110) and it's a breaking change, e.g. hamcrest uses it, which is arguably past its heyday, but still has more downloads than flixel (barely, but still).

**Comment by RealyUniqueName** [comment]
I thought it's necessary, but apparently [it's not](https://github.com/HaxeFoundation/haxe/pull/9672).
@Simn what do you think?

**Comment by Simn** [comment]
Many keywords are not strictly necessary for parsing. However, I prefer to have a rigid structure for syntax, especially at expression-level.

**Comment by Simn** [comment]
Having said that... there really isn't much benefit in making `is` a keyword right now. Unless we come across some parsing problem, we can just let it sit and then think about keywording it in Haxe 5.

Though in that case I suppose it would be nice if there was a way to disable the warning we're now getting...

**Comment by RealyUniqueName** [comment]
I think it's better to just revert the warning in that case.

**Comment by Simn** [comment]
I think we can close here.
